Mcast src improve

vomikodved opened the pull request:
This patch reviews sscpoe device add and config procedure.
First of all, it adds explicit local multicast search setup with src interface and mcast ttl selection.
There are several UX patches as well.

### New "add record" flow 
the updated flow for SSCPOE integration is described below

The user chooses:
├── Cloud →flow left intact
├──Local(Web) → set IP+password → Web access verification → add Web/cancel 
└── Local(Multicast) → Interface refinement/ttl → do search +Web access verification → selector: old API(all devices found) or Web(those of found with web access granted) → add selected/cancel

#### "Local(Web)" steps
- Enter IP+ password
- Access verification(HTTP login attempt) 
- showing a list of items(one item max in this case): Add WEB GS108..., S/N: ...(IP) 
- creating an entry

#### "Local(Multicast)" steps
- Interface selection + mcast TTL edit 
- Device search 
- HTTP-Availability check(with default password) 
- Showing the list of items: 
    + Add old API GS108..., S/N: ...(IP)
    + Add WEB GS108..., S/N: ...(IP) ← if Web access is successful
- Creating an entry for selected item


#### Technical note
The logic of local_select\ is universal
- on the path from web local, it has a maximum of one element in the list, and with multicast local, there can be several elements.
- form the list of local_devices with the service field of the method of addition,
- add a duplicate device with the web method to multicast_local upon successful call to check_web_availability(copy the current element and replace the addition method from multicast/old to web)
- interpret the list in local_select only, taking into account the addition method for each element; minimum calculations: displaying the entire array piecemeal, and calling the corresponding handler element

vomikodved committed:
protocol.py update to open bind_interface and/or mcast ttl customization

vomikodved committed:
manifest fix

vomikodved committed:
added interface flow for local discovery interface and mcast ttl explicit setting

vomikodved committed:
ttl selection form fix

vomikodved committed:
local multicast form UX clarified

vomikodved committed:
filter out web-inaccessible switches from available list after old-api detection

vomikodved committed:
update manifest date

vomikodved committed:
update manifest date

vomikodved committed:
fixed to config flow

slydiman commented on custom_components/sscpoe/protocol.py:
```diff
@@ -440,22 +443,92 @@ def get_host_ip():
+            raise ValueError(f"Invalid TTL={ttl}, expected 1..255")
+        sock.setsockopt(socket.IPPROTO_IP, socket.IP_MULTICAST_TTL, int(ttl))
+    else:
+        sock.setsockopt(socket.IPPROTO_IP, socket.IP_MULTICAST_TTL, 2)
```
int(ttl) is redundant.
```suggestion
    if ttl is None:
        ttl = 2
    if ttl < 1 or ttl > 255:
        raise ValueError(f"Invalid TTL={ttl}, expected 1..255")
    sock.setsockopt(socket.IPPROTO_IP, socket.IP_MULTICAST_TTL, ttl)
```

slydiman commented on custom_components/sscpoe/config_flow.py:
```diff
@@ -11,78 +15,92 @@
     local_devices: list[dict] = None
+    entry: ConfigEntry | None = None
+
+    def _extract_model_from_sn(self, sn: str | None) -> str:
```
Use SSCPOE_Coordinator.model_from_sn(sn) instead.
Note SN PS208GV3254201F6CGQ7******. Your code extracts `PS208GV325` which is wrong.

slydiman commented on custom_components/sscpoe/config_flow.py:
```diff
@@ -11,78 +15,92 @@
 from .protocol import (
+    SSCPOE_LOCAL_DEF_PASSWORD,
     SSCPOE_local_search,
+    _get_ipv4_for_interface,
```
If you import `_get_ipv4_for_interface` here, this function is not private. Please rename to `get_ipv4_for_interface` then.

Or even better move `_get_interface_choices` to protocol.py and rename to `get_interface_choices`

slydiman commented on custom_components/sscpoe/config_flow.py:
```diff
@@ -11,78 +15,92 @@
     async def async_step_local(self, user_input: dict[str, str] = None) -> FlowResult:
         errors: dict[str, str] = {}
+        sn = None
+        password = "123456"
```
Use `SSCPOE_LOCAL_DEF_PASSWORD` instead of `123456` here and below.

slydiman commented on custom_components/sscpoe/config_flow.py:
```diff
@@ -96,31 +114,34 @@ def login():
     async def async_step_cloud(self, user_input: dict[str, str] = None) -> FlowResult:
         errors: dict[str, str] = {}
+        email = None
+        password = "123456"
```
SSCPOE_LOCAL_DEF_PASSWORD

slydiman commented on custom_components/sscpoe/config_flow.py:
```diff
@@ -145,19 +166,281 @@ def login():
+        """Screen 2W: web credentials check, then go to local confirm screen."""
+        errors: dict[str, str] = {}
+        ip = None
+        password = "123456"
```
SSCPOE_LOCAL_DEF_PASSWORD

slydiman commented on custom_components/sscpoe/config_flow.py:
```diff
@@ -145,19 +166,281 @@ def login():
     async def async_step_web(self, user_input: dict[str, str] = None) -> FlowResult:
         errors: dict[str, str] = {}
+        ip = None
+        password = "123456"
```
SSCPOE_LOCAL_DEF_PASSWORD

slydiman commented on custom_components/sscpoe/config_flow.py:
```diff
@@ -145,19 +166,281 @@ def login():
+                    data_schema=vol.Schema(
+                        {
+                            vol.Required(CONF_IP_ADDRESS, default=ip): str,
+                            vol.Required(CONF_PASSWORD, default="123456"): str,
```
SSCPOE_LOCAL_DEF_PASSWORD

slydiman commented on custom_components/sscpoe/config_flow.py:
```diff
@@ -145,19 +166,281 @@ def login():
+                    data_schema=vol.Schema(
+                        {
+                            vol.Required(CONF_ID, default=sn): str,
+                            vol.Required(CONF_PASSWORD, default="123456"): str,
```
SSCPOE_LOCAL_DEF_PASSWORD

slydiman commented on custom_components/sscpoe/config_flow.py:
```diff
@@ -145,19 +166,281 @@ def login():
+        actions = {}
+        for device in (self.local_devices or []):
+            if device.get("add_method") == "old":
+                sn = device.get("sn") or device.get("ip", "unknown")
```
I don't see a reason to use IP here. Please replace with
sn = device.get("sn", "unknown")

slydiman commented on custom_components/sscpoe/config_flow.py:
```diff
@@ -145,19 +166,281 @@ def login():
+
+            elif device.get("add_method") == "web":
+                ip = device["ip"]
+                sn = device.get("sn") or f"web_{ip.replace('.', '_')}"
```
```suggestion
                sn = device.get("sn", "?")
```

slydiman commented on custom_components/sscpoe/config_flow.py:
```diff
@@ -145,19 +166,281 @@ def login():
+                ip = _get_ipv4_for_interface(name)
+                if ip:
+                    interfaces.append(f"{name} ({ip})")
+            return [DEFAULT_BIND_INTERFACE] + sorted(interfaces)
+        except Exception:
+            return [DEFAULT_BIND_INTERFACE]
```
```suggestion
        except Exception:
            pass
        return [DEFAULT_BIND_INTERFACE] + sorted(interfaces)            
```

slydiman commented on custom_components/sscpoe/config_flow.py:
```diff
@@ -145,19 +166,281 @@ def login():
+            )        
+        
+        bind_default = DEFAULT_BIND_INTERFACE
+        set_ttl_default = False
+        ttl_default = 2
```
```suggestion
        DEFAULT_TTL = 2
```
and move to the file beginning.

slydiman commented on custom_components/sscpoe/config_flow.py:
```diff
@@ -145,19 +166,281 @@ def login():
+            errors=errors,
+        )    
+    
+    async def _check_web_accessibility(self, devices: list[dict]) -> None:
+        """Check HTTP management accessibility with default password for devices."""
+        
+        for device in devices:
+            ip = device.get("ip")
+            if not ip:
+                continue
+                
+            try:
+                # Use the same method as in async_step_local_web
+                err, token = await self.hass.async_add_executor_job(
+                    SSCPOE_web_login, ip, SSCPOE_LOCAL_DEF_PASSWORD
+                )
+                if not err:
+                    # Add token for further use
+                    device["web_token"] = token
+                else:
+                    LOGGER.debug("SSCPOE HTTP access check failed for %s: %s", ip, err)
+            except Exception as e:
+                LOGGER.debug("SSCPOE HTTP access check exception for %s: %s", ip, e)
+                
+        return None
```
Move it as check_web_accessibility to protocol.py
Note it is bad idea to try login with a default password.
Use web request cmd=100 instead w/o a token.
```suggestion
    async def check_web_accessibility(self, device: dict) -> bool:
        """Check HTTP management accessibility."""
        ip = device.get("ip")
        if not ip:
            return False
        try:
            j, err = await self.hass.async_add_executor_job(
                SSCPOE_web_request, ip, None, 100
            )
            if j is not None:
                return True
            else:
                LOGGER.debug(f"SSCPOE WEB access check failed for {ip}: {err}")
        except Exception as e:
            LOGGER.debug(f"SSCPOE WEB access check exception for {ip}: {str(e)}")
        return False
```

slydiman commented on custom_components/sscpoe/config_flow.py:
```diff
@@ -145,19 +166,281 @@ def login():
+
+                    return await self.async_step_local_select()
+
+            bind_default = bind_iface
+            ttl_default = ttl
```
Any purpose of these lines?
```suggestion
```

slydiman commented on custom_components/sscpoe/config_flow.py:
```diff
@@ -145,19 +166,281 @@ def login():
+            ttl_default = ttl
+    
+        schema_dict: dict = {
+            vol.Required(CONF_BIND_INTERFACE, default=bind_default): vol.In(iface_choices),
```
Use DEFAULT_BIND_INTERFACE instead of bind_default. Just remove bind_default at all.

slydiman commented on custom_components/sscpoe/config_flow.py:
```diff
@@ -145,19 +166,281 @@ def login():
+                errors[CONF_IP_ADDRESS] = "invalid_ip"
+            else:
+    
+                devices = [{"ip": ip}]
```
Just return the original behavior here. 
You cannot try to login here with the default password. The device may be already activated with user's password. Your client IP may be banned in case of a wrong password. Don't check WEB access. Just try to login with the provided IP / password and add the entry immediately. There is no reason to show the device SN, FW and IP after the correct login and wait for the Submit button.

slydiman commented on custom_components/sscpoe/protocol.py:
```diff
@@ -429,6 +429,9 @@ def SSCPOE_local_syn():
 host_ip = None
 
+# fallback defaults (set to None / "default" for normal behaviour)
+bind_interface_default = None   # e.g. "enp0s6"
+mcast_ttl_default = None        # set None to keep current hardcoded default in code, or an int 1..255
```
`bind_interface_default` and `mcast_ttl_default` are useless.
You must save the user's values in the device's enrty and pass them to SSCPOE_local_send().

slydiman commented on custom_components/sscpoe/config_flow.py:
```diff
@@ -11,78 +15,92 @@
+        sn = device.get("sn")
+
+        # Always extract model from SN for consistency, even if 'model' exists
+        device["model"] = self._extract_model_from_sn(sn)
```
Don't replace the existing model. The model may be `GS105` and `GS105V1_1`. Note SN does not contain `_` symbol (for example `GS105V1252***`), but the correct model printed on the label may contain `_`.

slydiman commented:
I don't see how to enter the password using `Local (Multicast)` method. 
Note the device may be already activated with non default password.

slydiman commented:
@vomikodved
I have pushed [the commit](https://github.com/slydiman/sscpoe/commit/634366b9bee171a165dafc43d717c89f47af6213) based on this PR
- Added network interface and TTL selection for local multicast protocol
- Added checking for WEB protocol accessibility in auto search mode
- Refactored setup flow

The interface and TTL are stored in the device enrty now.
I have added to protocol.py the following methods: SSCPOE_web_accessibility, SSCPOE_normalize_device, SSCPOE_get_interfaces, get_if_ip, bind_if
and moved SSCPOE_model_from_sn from coordinator.py

Thanks for your effort.

Feel free to add the translation JSON.
We can discuss the setup flow.

vomikodved commented:
> @vomikodved I have pushed [the commit](https://github.com/slydiman/sscpoe/commit/634366b9bee171a165dafc43d717c89f47af6213) based on this PR
> 
> * Added network interface and TTL selection for local multicast protocol
> * Added checking for WEB protocol accessibility in auto search mode
> * Refactored setup flow
> 
> The interface and TTL are stored in the device enrty now. I have added to protocol.py the following methods: SSCPOE_web_accessibility, SSCPOE_normalize_device, SSCPOE_get_interfaces, get_if_ip, bind_if and moved SSCPOE_model_from_sn from coordinator.py
> 
> Thanks for your effort.
> 
> Feel free to add the translation JSON. We can discuss the setup flow.

thanks, I'll try the version this evening or tomorrow

vomikodved commented:
> I don't see how to enter the password using `Local (Multicast)` method. Note the device may be already activated with non default password.

should we add password field to this (mcast discovery setup) form?
<img width="412" height="317" alt="image" src="https://github.com/user-attachments/assets/08c16d79-ae68-4235-aa8e-031444a27c65" />

if already discovered, there _is_ such a field in both web and mcast addition forms:
<img width="412" height="317" alt="image" src="https://github.com/user-attachments/assets/d15098e4-f069-4cef-9ecf-7835d48d9f4f" />
<img width="412" height="317" alt="image" src="https://github.com/user-attachments/assets/c4e3acec-e0b4-43d3-920a-d41d27d74e98" />

slydiman commented:
> should we add password field to this (mcast discovery setup) form?

No. We can find few devices with different passwords. 
We must ask the password only after selecting the device.

vomikodved commented:
> @vomikodved I have pushed [the commit](https://github.com/slydiman/sscpoe/commit/634366b9bee171a165dafc43d717c89f47af6213) based on this PR
> 
> * Added network interface and TTL selection for local multicast protocol
> * Added checking for WEB protocol accessibility in auto search mode
> * Refactored setup flow
> 
> The interface and TTL are stored in the device enrty now. I have added to protocol.py the following methods: SSCPOE_web_accessibility, SSCPOE_normalize_device, SSCPOE_get_interfaces, get_if_ip, bind_if and moved SSCPOE_model_from_sn from coordinator.py
> 
> Thanks for your effort.
> 
> Feel free to add the translation JSON. We can discuss the setup flow.



> @vomikodved I have pushed [the commit](https://github.com/slydiman/sscpoe/commit/634366b9bee171a165dafc43d717c89f47af6213) based on this PR
> 
> * Added network interface and TTL selection for local multicast protocol
> * Added checking for WEB protocol accessibility in auto search mode
> * Refactored setup flow
> 
> The interface and TTL are stored in the device enrty now. I have added to protocol.py the following methods: SSCPOE_web_accessibility, SSCPOE_normalize_device, SSCPOE_get_interfaces, get_if_ip, bind_if and moved SSCPOE_model_from_sn from coordinator.py
> 
> Thanks for your effort.
> 
> Feel free to add the translation JSON. We can discuss the setup flow.

thank you, the scenario succeeds for my 3 switches in any combinations. 
the  only small thing this setup flow lacks is unified spec of HA object creation (multicast-controlled devices are created using serial in the record name, while web-controlled devices have an IP address). But that's really minor disadvantage.

vomikodved committed:
Merge branch 'main' into mcast-src-improve

slydiman commented:
Ok. I will add translations/ru.json and release the new version. 
We can just close the PR then. 
Thanks for interface enumerating/binding implementations.

vomikodved commented:
2025.12.21 release implements ideas of this PR. thus, it's not needed anymore, closing
